pkg/http/server_test.go

```diff
@@ -98,79 +99,85 @@ func TestNewServer(t *testing.T) {
+		tls.TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,    // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+		tls.TLS_ECDHE_ECDSA_WITH_RC4_128_SHA,        // Broken cipher
+		tls.TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA,     // Sweet32
+		tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA,      // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+		tls.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,   // No countermeasures against timing attacks
+		tls.TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,      // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+		tls.TLS_ECDHE_RSA_WITH_RC4_128_SHA,          // Broken cipher
+
+		// all RSA-PKCS1-v1.5 ciphers are disabled - danger of Bleichenbacher attack variants
+		tls.TLS_RSA_WITH_3DES_EDE_CBC_SHA,   // Sweet32
+		tls.TLS_RSA_WITH_AES_128_CBC_SHA,    // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+		tls.TLS_RSA_WITH_AES_128_CBC_SHA256, // No countermeasures against timing attacks
+		tls.TLS_RSA_WITH_AES_256_CBC_SHA,    // Go stack contains (some) countermeasures against timing attacks (Lucky13)
+		tls.TLS_RSA_WITH_RC4_128_SHA,        // Broken cipher
+
+		tls.TLS_RSA_WITH_AES_128_GCM_SHA256, // Disabled because of RSA-PKCS1-v1.5 - AES-GCM is considered secure.
+		tls.TLS_RSA_WITH_AES_256_GCM_SHA384, // Disabled because of RSA-PKCS1-v1.5 - AES-GCM is considered secure.
+	}
 
-// Tests the server accepts only TLS connections when at least
-// one of the supported (secure) cipher suites is supported by the client.
-func TestTLSCiphers(t *testing.T) {
 	certificate, err := getTLSCert()
 	if err != nil {
 		t.Fatalf("Unable to parse private/certificate data. %v\n", err)
 	}
-	port := getNextPort()
-	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(http.StatusOK)
-		w.Write(nil)
-	})
-	server := NewServer([]string{"127.0.0.1:" + port}, handler, &certificate)
-	go func() { server.Start() }()
-	defer server.Shutdown()
 
-	for !server.IsStarted() { // block until server is available
+	testCases := []struct {
+		ciphers            []uint16
+		resetServerCiphers bool
+		expectErr          bool
+	}{
+		{nil, false, false},
+		{defaultCipherSuites, false, false},
+		{unsupportedCipherSuites, false, true},
+		{nil, true, false},
+		{defaultCipherSuites, true, false},
+		{unsupportedCipherSuites, true, false},
 	}
-	for i, testCase := range tlsCipherSuitesTests {
-		client := http.Client{
-			Transport: &http.Transport{
-				TLSClientConfig: &tls.Config{
-					InsecureSkipVerify: true,
-					CipherSuites:       testCase.ciphers,
+
+	for i, testCase := range testCases {
+		func() {
+			addr := "127.0.0.1:" + getNextPort()
+
+			server := NewServer([]string{addr},
+				http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+					fmt.Fprintf(w, "Hello, world")
+				}),
+				&certificate)
+			if testCase.resetServerCiphers {
+				// Use Go default ciphers.
+				server.TLSConfig.CipherSuites = nil
```

Comment:
Can you explain what you're trying to achieve with this?
Handling `TLSConfig.CipherSuites` correctly is the responsibility of the standard library. Testing the standard library is IMO not useful. So setting `TLSConfig.CipherSuites` to an arbitrary value should not change anything except that a client has to support at least one of those cipher suites.

 So there are two important cases:
1. A client who only supports "unsecure" ciphers should not be able to connect
2. A client who supports at least one "secure" cipher should be able to connect
(3. A default Go client should be able to connect for backward comparability)

What is the point of testing whether a default Go client can connect to a server with the default cipher suites?

Reply:
Could you list down what test cases are valid so that I can remove rest of them?

Reply:
- [That](https://github.com/aead/minio/pull/3/files#diff-e3962e3c7025eecdb6dc51672e35b223R135) would be the backward compatibility test (default Go client can connect)
- [That](https://github.com/aead/minio/pull/3/files#diff-e3962e3c7025eecdb6dc51672e35b223R136) would test whether a client which supports (at least one) cipher suites of the server can connect
- [That](https://github.com/aead/minio/pull/3/files#diff-e3962e3c7025eecdb6dc51672e35b223R137) would test that the server rejects clients who does not support secure ciphers - and following that the server does not support "insecure" ciphers.

The other 3 are not really useful IMO - Or am I missing your point?

Reply:
To cover with your current test code, I am going to retain https://github.com/aead/minio/pull/3/files#diff-e3962e3c7025eecdb6dc51672e35b223R138 and remove other two

Reply:
Okay, lets agree on that 👍